File: sidequest/backend/schemas/events.py
"""Typed SSE event payloads.

This module is the backend half of the protocol contract. It must stay
mirrored with frontend/types/events.ts (API_SPEC.md section 2). The schema
snapshot test in tests/test_events.py guards the wire shapes.
"""

from typing import Literal

from pydantic import BaseModel

from schemas.intent import IntentPublic

ArtifactStage = Literal["planning", "generating", "verifying", "revising", "postprocessing"]


class MetaPayload(BaseModel):
    intent: IntentPublic
    canonical_concept: str
    cache: Literal["hit", "miss"]


class TextDeltaPayload(BaseModel):
    chunk: str


class TextDonePayload(BaseModel):
    pass


class ArtifactStatusPayload(BaseModel):
    stage: ArtifactStage


class ArtifactDeltaPayload(BaseModel):
    chunk: str


class ArtifactDonePayload(BaseModel):
    artifact_id: str
    title: str
    html: str


class ArtifactFailedPayload(BaseModel):
    reason: str
    detail_user: str
    retryable: bool


class TutorMsgPayload(BaseModel):
    text: str


class Usage(BaseModel):
    tokens_in: int = 0
    tokens_out: int = 0
    cost_usd: float = 0.0


class Timings(BaseModel):
    first_token: int = 0
    artifact_total: int = 0


class DonePayload(BaseModel):
    usage: Usage
    timings_ms: Timings


# Wire name for each payload type. `meta` is always first, `done` always last.
EVENT_NAMES: dict[type[BaseModel], str] = {
    MetaPayload: "meta",
    TextDeltaPayload: "text_delta",
    TextDonePayload: "text_done",
    ArtifactStatusPayload: "artifact_status",
    ArtifactDeltaPayload: "artifact_delta",
    ArtifactDonePayload: "artifact_done",
    ArtifactFailedPayload: "artifact_failed",
    TutorMsgPayload: "tutor_msg",
    DonePayload: "done",
}


def event_name(payload: BaseModel) -> str:
    try:
        return EVENT_NAMES[type(payload)]
    except KeyError as exc:  # pragma: no cover - programming error, not runtime input
        raise TypeError(
            f"{type(payload).__name__} is not a registered SSE payload"
        ) from exc


def format_sse(payload: BaseModel, event_id: int) -> str:
    """Encode one event as an SSE frame with a monotonic id for future resume."""
    data = payload.model_dump_json()
    return f"id: {event_id}\nevent: {event_name(payload)}\ndata: {data}\n\n"

File: sidequest/frontend/types/events.ts
/**
 * Typed SSE event schemas. Mirrors backend/schemas/events.py exactly
 * (API_SPEC.md section 2). If you change one side, change the other in
 * the same commit; the backend snapshot tests guard the wire shapes.
 */
import { z } from "zod";

export const IntentSchema = z.object({
  artifact_type: z.enum([
    "simulation",
    "explorable_diagram",
    "virtual_experiment",
    "data_visualization",
    "text_only",
  ]),
  domain: z.enum(["physics", "chemistry", "biology", "earth_space", "math_adjacent"]),
  complexity: z.union([z.literal(1), z.literal(2), z.literal(3)]),
});

export const MetaSchema = z.object({
  intent: IntentSchema,
  canonical_concept: z.string(),
  cache: z.enum(["hit", "miss"]),
});

export const TextDeltaSchema = z.object({ chunk: z.string() });
export const TextDoneSchema = z.object({});

export const ArtifactStatusSchema = z.object({
  stage: z.enum(["planning", "generating", "verifying", "revising", "postprocessing"]),
});
export const ArtifactDeltaSchema = z.object({ chunk: z.string() });
export const ArtifactDoneSchema = z.object({
  artifact_id: z.string(),
  title: z.string(),
  html: z.string(),
});
export const ArtifactFailedSchema = z.object({
  reason: z.string(),
  detail_user: z.string(),
  retryable: z.boolean(),
});

export const TutorMsgSchema = z.object({ text: z.string() });

export const DoneSchema = z.object({
  usage: z.object({
    tokens_in: z.number(),
    tokens_out: z.number(),
    cost_usd: z.number(),
  }),
  timings_ms: z.object({
    first_token: z.number(),
    artifact_total: z.number(),
  }),
});

export const eventSchemas = {
  meta: MetaSchema,
  text_delta: TextDeltaSchema,
  text_done: TextDoneSchema,
  artifact_status: ArtifactStatusSchema,
  artifact_delta: ArtifactDeltaSchema,
  artifact_done: ArtifactDoneSchema,
  artifact_failed: ArtifactFailedSchema,
  tutor_msg: TutorMsgSchema,
  done: DoneSchema,
} as const;

export type EventName = keyof typeof eventSchemas;

export type Meta = z.infer<typeof MetaSchema>;
export type Intent = z.infer<typeof IntentSchema>;

export type AxiomEvent = {
  [K in EventName]: { type: K; data: z.infer<(typeof eventSchemas)[K]> };
}[EventName];

/** Validate one wire event. Unknown names and invalid payloads return null
 *  (callers drop them silently and count, per API_SPEC bridge/listener rules). */
export function parseEvent(name: string, data: unknown): AxiomEvent | null {
  const schema = (eventSchemas as Record<string, z.ZodTypeAny>)[name];
  if (!schema) return null;
  const result = schema.safeParse(data);
  if (!result.success) return null;
  return { type: name, data: result.data } as AxiomEvent;
}
